src/main/scala/ch/epfl/bluebrain/nexus/iam/realms/Groups.scala

```diff
@@ -0,0 +1,182 @@
+package ch.epfl.bluebrain.nexus.iam.realms
+
+import java.time.Instant
+import java.util.concurrent.TimeUnit
+
+import akka.actor.{Actor, ActorLogging, ActorRef, ActorSystem, PoisonPill, Props, ReceiveTimeout}
+import akka.cluster.sharding.ShardRegion.{ExtractEntityId, ExtractShardId, Passivate}
+import akka.cluster.sharding.{ClusterSharding, ClusterShardingSettings}
+import akka.http.scaladsl.client.RequestBuilding._
+import akka.http.scaladsl.model.headers.OAuth2BearerToken
+import akka.pattern.{AskTimeoutException, ask}
+import akka.util.Timeout
+import cats.effect.{Async, IO, Timer}
+import cats.implicits._
+import ch.epfl.bluebrain.nexus.commons.http.HttpClient
+import ch.epfl.bluebrain.nexus.iam.auth.AccessToken
+import ch.epfl.bluebrain.nexus.iam.config.AppConfig.GroupsConfig
+import ch.epfl.bluebrain.nexus.iam.realms.Groups._
+import ch.epfl.bluebrain.nexus.iam.types.IamError.{InternalError, OperationTimedOut}
+import ch.epfl.bluebrain.nexus.iam.types.Identity.Group
+import ch.epfl.bluebrain.nexus.iam.types.Label
+import ch.epfl.bluebrain.nexus.sourcing.akka.Retry
+import ch.epfl.bluebrain.nexus.sourcing.akka.syntax._
+import com.nimbusds.jwt.JWTClaimsSet
+import io.circe.{Decoder, HCursor, Json}
+import journal.Logger
+
+import scala.concurrent.duration.Duration
+import scala.reflect.ClassTag
+import scala.util.Try
+import scala.util.control.NonFatal
+
+/**
+  * Extracts and caches caller group set using the access token or the realm user info endpoint as sources.
+  *
+  * @param ref the underlying cluster sharding actor ref proxy
+  */
+class Groups[F[_]](ref: ActorRef)(implicit cfg: GroupsConfig, hc: HttpClient[F, Json], F: Async[F], T: Timer[F]) {
```

Comment:
Do you need Async[F] and Timer[F]? Wouldn't Effect[F] be enough?

Reply:
Timer is required for Retry and Async is enough (no need for Effect)